### DISCUSSION TITLE
for parallel tests, how does the service handle multiple requests to the same endpoint at the same time

### DISCUSSION THREAD

**Discussion by ponienchen** [sorry-cypress/sorry-cypress#330] [Q&A] [answered]
As the subject says, when there are multiple cypress instances asking the director service for a new task, how are those simultaneous requests handled by the service ?

Do those requests get placed in a queue somewhere by the director service, which then responds to each request one by one?

Can someone kindly help me understand this part ?

**Comment by agoldis** [comment, marked as answer]
There's no queue per-se. Behind the scenes it uses Mongo [atomicity](https://docs.mongodb.com/manual/core/write-operations-atomicity/#atomicity) to avoid concurrency issues. Here's the [code](https://github.com/sorry-cypress/sorry-cypress/blob/b2962892c743219e43fdf289617d73b20dd06b2f/packages/director/src/execution/mongo/runs/run.model.ts#L103)

**Reply by ponienchen** [reply]
Got it.
As for the in-memory.ts, can you help me understand how it achieves atomicity ?
Does the node event loop play a part in this ?

**Reply by agoldis** [reply]
in-memory update of claimed spec isn't async i.e. node loop isn't interrupted (at least isn't supposed to)

**Reply by ponienchen** [reply]
got it!
Thanks for helping to clarify this for me!
